docs/proposals/flexvolume-api.md

```diff
@@ -0,0 +1,187 @@
+
+```protobuf
+
+
+// Capabilities of the driver
+message DriverCapabilities {
+    enum ReadWriteMode {
+        READ_ONLY = 0;
+        READ_WRITE_ONCE = 1;
+        READ_WRITE_MANY = 2;
+    }
+	ReadWriteMode supported_read_write_mode = 1;
+    enum AttachmentPolicy {
+        NONE = 0;  // Driver does not support attach/detach.
+        LOCAL = 1; // Driver supports local volume attachments from kubelet.
+        REMOTE = 2; // Driver supports remote volume attachments to a node from controller-manager.
+    }
+    AttachmentPolicy supported_attachment_policy = 2;
+    enum Capabilities {
+        DYNAMIC_PROVISIONING = 0;
+        CUSTOM_MOUNT = 1;
+        SELINUX = 2,
+        OWNERSHIP_MANAGEMENT = 3;
+        METRICS = 4;
+    }
+    repeated Capabilities supported_capabilities = 3;
+}
+
+message ProbeDriverRequest {}
+
+message ProbeDriverResponse {
+    DriverCapabilities capabilities = 1;
+    repeated string options = 2;
+}
+
+service FlexVolumeService {
+    rpc ProbeDriver(ProbeDriverRequest) returns (ProbeDriverResponse) {}
+    rpc Create(CreateRequest) returns (CreateResponse) {}
+    rpc Delete(DeleteRequest) returns (DeleteResponse) {}
+    rpc Attach(AttachRequest) returns (AttachResponse) {}
+    rpc Detach(DetachRequest) returns (DetachResponse) {}
+    rpc Mount(MountRequest) returns (MountResponse) {}
+    rpc Unmount(UnmountRequest) returns (UnmountResponse) {}
+    rpc GetMetrics(MetricsRequest) returns (MetricsResponse) {}
+}
+
+// Volume source.
+message Spec {
+    string name = 1; // Name of the volume.
+    string fstype = 2; // Requested file system type.
+    string read_write_mode = 3;
+    uint64 size = 4; // Size of the volume.
+    map<string, string> secrets = 5; // Secrets required for the driver to talk to its own controller.
+    map<string, string> options = 6; // Extra options passed to the driver.
+}
+
+message CreateRequest{
+    Spec spec = 1; // Volume specification.
+}
+
+message CreateResponse{
+    string id = 1; // Volume identifier, UUID/name
+}
+
+message DeleteRequest{
+    string name = 1; // Name of the volume.
+}
+
+message DeleteResponse{}
+
+message AttachRequest {
+    Spec spec = 1; // Volume spec.
+    string host = 2; // Name of the host to attach volume to.
+}
+
+message AttachResponse{
+    string device_path = 1; // Path to the device where the volume is attached.
+}
+
+message DetachRequest {
+    string name = 1; // Name of the volume.
+    string host = 2; // Name of the host to detach volume from.
+    bool sync = 3; // Wait for detach to finish.
+    map<string, string> secrets = 4; // Extra options passed to the driver.
+}
+
+message DetachResponse {}
+
+message MountRequest {
+    Spec spec = 1; // Volume specifciation.
+    string mount_path = 2; // Recommended path for the driver to mount the volume.
+}
+
+message MountResponse {
+    string mount_path = 1; // Path where the driver actually mounted the volume.
+}
+
+message UnmountRequest {
+    string mount_path = 1;
```

Comment:
I think passing whole Spec won't harm here if it is known, a driver might be interested e.g. in secrets.